dwave/preprocessing/src/presolveimpl.hpp

```diff
@@ -322,10 +453,282 @@ class PresolverImpl {
         return model_.restore(std::move(reduced));
     }
 
+    /// Clear redundant constraints by turning them into 0 == 0 constraints.
+    /// We don't actually remove them (yet) because we don't want to reallocate
+    /// our constraint vector.
+    bool technique_clear_redundant_constraint(constraint_type& constraint) {
+        // not currently implemented for quadratic constraints.
+        // todo: extend to quadratic
+        if (!constraint.is_linear()) {
+            return false;
```

Comment:
Having so many different return points in a function/method is sort of an anti-pattern...

Reply:
We could instead carry around a bunch of bools, or nest the function even deeper into `if` statements. But IMO that would make it even less readable.

Reply:
Having so many returns is generally frowned upon, if not actually considered bad practice or an anti-pattern.  Partially, this is because it can make debugging a lot more difficult, especially as the anti-pattern replicateds itself throughout the code-base.  It also makes the code less self-documenting, since each case needs to be teased apart.  Functions like this can also become a maintenance nightmare as the various exit-points can cause unexpected side-effects when an edit is made a year or two in the future.

Obviously, there are always exceptions and counter-examples but they should be that: exceptional.

Reply:
Sure, I agree in general that multiple returns are to be avoided when possible. I am just genuinely unsure how I would refactor this particular method to make that happen without obfuscating the code even more. I attempted to mitigate it somewhat with comments, and I think https://github.com/dwavesystems/dimod/issues/1330 will help a bit by making the various if branches a bit simpler.

I'll take another crack at refactoring it, but it's not obvious to me how I would proceed, especially if I only want to calculate `minimal_activation` and `maximal_activation` once. Suggestions welcome.

Reply:
Ok, I took a second look. I didn't end up changing the pattern, though I am heartened by the LLVM Coding Standard's [opinion](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code) on the matter. On the up side, I did find a bug with the existing implementation and fixed it. That, combined with https://github.com/dwavesystems/dimod/issues/1330 makes the method clearer now I think.